consistency around terminology random (#318)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

For each claim that is an object property and that is to be made selectively disclosable, the Issuer MUST create a Disclosure as follows:

 * Create an array of three elements in this order:
   1. A salt value. MUST be a string. See (#salt-entropy) and (#salt_minlength) for security considerations. It is RECOMMENDED to base64url-encode minimum 128 bits of cryptographically secure random data, producing a string. The salt value MUST be unique for each claim that is to be selectively disclosed. The Issuer MUST NOT disclose the salt value to any party other than the Holder.

>    2. The claim name, or key, as it would be used in a regular JWS payload. The value MUST be a string.

Reviewer: IMHO JWT payload is fine and conveys more meaning. 
```suggestion
   2. The claim name, or key, as it would be used in a regular JWT payload. The value MUST be a string.
```

Reviewer: a JWS payload "can contain an arbitrary sequence of octets" while JWT conveys JSON and claims

Author: yeah. I agree that JWT payload is pretty well understood.

>    3. The claim value, as it would be used in a regular JWS payload. The value MAY be of any type that is allowed in JSON, including numbers, strings, booleans, arrays, and objects.

Reviewer: ```suggestion
   3. The claim value, as it would be used in a regular JWT payload. The value MAY be of any type that is allowed in JSON, including numbers, strings, booleans, arrays, and objects.
```

associated with the Holder, or a reference thereto.

It is out of the scope of this document to describe how the Holder key pair is

> established. For example, the Holder MAY create a key pair and provide a public key to the Issuer,

Author: feedback received: 
> "It is out of the scope of this document to describe how the Holder key pair is established. For example, the Holder MAY provide a key pair to the Issuer,…" Requiring the holder to provide also the private key to the issuer appears not to be a good example.

suggestion was "the Holder MAY provide a public key to the Issuer and prove possession of the corresponding private key" but I slightly played with the words.

> * in the JOSE header,

Reviewer: ```suggestion
* in the JWS header,
```

Author: per RFC, it's JOSE header.. https://www.rfc-editor.org/rfc/rfc7515#section-2
per DPoP/convention, JWT header might be ok too..

Reviewer: yeah, probably any of them would be ok actually

Author: ```suggestion
* in the JWT header,
```

Author: there are other instances where we use `JOSE header`. so let's decide in this issue comment on the term and i will change everywhere accordignly, if needed.

Reviewer: let's just keep `JOSE header`

    * `typ`: REQUIRED. MUST be `kb+jwt`, which explicitly types the Key Binding JWT as recommended in Section 3.11 of [@!RFC8725].
    * `alg`: REQUIRED. A digital signature algorithm identifier such as per IANA "JSON Web Signature and Encryption Algorithms" registry. MUST NOT be `none` or an identifier for a symmetric algorithm (MAC).

> * in the JWS payload,

Reviewer: or this would be okay by me too 
```suggestion
* in the JWT payload,
```

> * in the JWS payload,

Author: yeah, let's stick to `JWT payload`.

An SD-JWT is composed of the following:

* the Issuer-signed JWT

> * optionally the Disclosures

Author: this one might be controversial...

original feedback received was:

> According to clause 5.1 the Disclosures and the Key Binding JWT are both optional in an SD-JWT.
>
> Clause 5.11 explicitly specifies the serialized format of an SD-JWT if the Key Binding JWT is absent:
"In the case that there is no Key Binding JWT, the last
element MUST be an empty string and the last separating tilde character MUST NOT be omitted."
>
> Clause 5.11 does not explicitly specify the serialized format of an SD-JWT if the Disclosures are absent.
>
> Please specify explicitly the serialized format of an SD-JWT if the Disclosures are absent.
>
> Please indicate in the item list ("An SD-JWT is composed of the following…") that the Disclosures are optional (as already done for the Key Binding JWT).

Reviewer: I'd interpreted "The Disclosures" to mean zero or more. So to me the optionality was implied.  But this seems fine too.

Reviewer 2: ```suggestion
* zero or more Disclosures
```

Reviewer: I like the Dr.'s suggestion above of "zero or more Disclosures"

For such use cases, the SD-JWT SHOULD be transported as a single string. Key Binding MAY be achieved by signing the envelope JWT instead of including a separate Key Binding JWT in the SD-JWT.

> The following non-normative example shows an SD-JWT Presentation enveloped in a JWS payload:

Reviewer: I don't totally love the suggested wording below but I think it's a "more correct" explanation of the content of the example. The text that was there previously doesn't seem quite right and JWT->JWS makes it maybe less correct. 

```suggestion
The following non-normative example shows the claims of an SD-JWT Presentation enveloped in a JWT:
```

Author: maybe make it simpler? `the claims of an SD-JWT presentation` took me a bit of time to understand.

```suggestion
The following non-normative example shows an SD-JWT Presentation enveloped in a JWT:
```

Reviewer: That works. It's really showing the claims/payload of that JWT in which the SD-JWT is enveloped. But that's hard to describe. And what I'd suggested is pretty bad now that I read it again.

Here, the SD-JWT is shown as the value of an `_sd_jwt` claim where `eyJhbGci...emhlaUJhZzBZ` represents the Issuer-signed JWT and `eyJhb...dYALCGg` represents a Disclosure. The SD-JWT does not contain a Key Binding JWT as the outer container can be signed instead.

> Other specifications or profiles of this specification may define alternative formats for transporting an SD-JWT that envelope multiple such objects into one object, and provide Key Binding using means other than the Key Binding JWT.

Reviewer 2: ```suggestion
Other specifications or profiles of this specification may define alternative formats for transporting an SD-JWT that envelope multiple such objects into one object and provide Key Binding using means other than the Key Binding JWT.
```
